File: api/routes/jobs.py
"""Job status and SSE streaming routes."""

import asyncio
import json
from fastapi import APIRouter, HTTPException
from fastapi.responses import StreamingResponse

from api.routes.posters import jobs

router = APIRouter()


@router.get("/{job_id}/stream")
async def stream_job_progress(job_id: str):
    """Server-Sent Events stream for job progress updates."""
    if job_id not in jobs:
        raise HTTPException(status_code=404, detail="Job not found")

    async def event_generator():
        last_progress = -1
        last_status = None

        while True:
            if job_id not in jobs:
                yield f"event: error\ndata: {json.dumps({'message': 'Job not found'})}\n\n"
                break

            job = jobs[job_id]
            current_progress = job["progress"]
            current_status = job["status"]

            # Only send update if something changed
            if current_progress != last_progress or current_status != last_status:
                last_progress = current_progress
                last_status = current_status

                if current_status == "completed":
                    yield f"event: completed\ndata: {json.dumps({'download_url': job['download_url']})}\n\n"
                    break
                elif current_status == "failed":
                    yield f"event: error\ndata: {json.dumps({'message': job.get('error', 'Unknown error')})}\n\n"
                    break
                else:
                    yield f"event: progress\ndata: {json.dumps({'step': job['current_step'], 'percent': int(current_progress * 100), 'message': job['message']})}\n\n"

            await asyncio.sleep(0.5)

    return StreamingResponse(
        event_generator(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
            "X-Accel-Buffering": "no"
        }
    )

File: maptoposter-rs/src/api/handlers/jobs.rs
use std::sync::Arc;
use std::time::Duration;

use axum::{
    extract::{Path, State},
    response::sse::{Event, Sse},
};
use futures::stream::Stream;
use tokio_stream::wrappers::IntervalStream;
use tokio_stream::StreamExt;
use uuid::Uuid;

use crate::api::models::{JobStatus, ProgressUpdate};
use crate::api::state::AppState;
use crate::error::{AppError, Result};

/// Stream job progress updates via Server-Sent Events
pub async fn stream_progress(
    State(state): State<Arc<AppState>>,
    Path(job_id): Path<String>,
) -> Result<Sse<impl Stream<Item = std::result::Result<Event, std::convert::Infallible>>>> {
    let uuid = Uuid::parse_str(&job_id).map_err(|_| AppError::JobNotFound(job_id.clone()))?;

    // Verify job exists
    if state.get_job(uuid).is_none() {
        return Err(AppError::JobNotFound(job_id));
    }

    // Create an interval stream that polls job status
    let interval = tokio::time::interval(Duration::from_millis(500));
    let stream = IntervalStream::new(interval);

    let state_clone = state.clone();
    let stream = stream
        .map(move |_| {
            let job = state_clone.get_job(uuid);

            match job {
                Some(job) => {
                    let update = ProgressUpdate {
                        job_id: job.id.to_string(),
                        status: job.status,
                        progress: job.progress,
                        step: job.current_step.unwrap_or_default(),
                        message: job.message.unwrap_or_default(),
                    };

                    let data = serde_json::to_string(&update).unwrap_or_default();

                    // Check if job is terminal (completed or failed)
                    let is_terminal =
                        job.status == JobStatus::Completed || job.status == JobStatus::Failed;

                    if is_terminal {
                        // Send final event and signal end
                        Event::default().data(data).event("completed")
                    } else {
                        Event::default().data(data).event("progress")
                    }
                }
                None => {
                    // Job no longer exists
                    Event::default()
                        .data("{\"error\": \"Job not found\"}")
                        .event("error")
                }
            }
        })
        // Note: SSE will continue until client disconnects
        // The "complete" event tells client to close connection
        .map(Ok);

    Ok(Sse::new(stream).keep_alive(
        axum::response::sse::KeepAlive::new()
            .interval(Duration::from_secs(15))
            .text("keep-alive"),
    ))
}
